File: rules/nocertificate.js
const fs = require('fs');
const vscode = require('vscode');

var smell = {
    detect: (fileName, token) => {
        try{
            if (token.hasOwnProperty("line")) var lineno = token.line;
            if (token.hasOwnProperty("type")) var tokenType = token.type;
            
            const MSG = 'TLS verification has been skipped';
            const WARNING_MSG = MSG+' at line '+ lineno;
            const WARNING_MSG_ON_RETURN = token.hasOwnProperty("returnLine") ? WARNING_MSG+ token.returnLine : null;
            
            const contextVars = ['requests.Session.verify'];
            const httpLibs = ['requests.get','requests.Session.get', 'requests.post', 'requests.Session.get'];
            
            if(tokenType == 'variable' && contextVars.includes(token.name) && token.value == false) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG); 
            if(tokenType == "variable" && token.hasOwnProperty('valueSrc') && token.hasOwnProperty('funcKeywords')) {
                if (httpLibs.includes(token.valueSrc.toLowerCase())) 
                    for(const keyword of token.funcKeywords){
                        if (keyword[0] == 'verify' && keyword[1] == false) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                    }
            }
            if(tokenType == "function_call" && token.hasOwnProperty('name') && token.hasOwnProperty('keywords')) {
                if (httpLibs.includes(token.name.toLowerCase())){
                    token.keywords.map(keyword => {
                        if (keyword[0] == 'verify' && keyword[1] == false) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                    });
                }
            }
            if(tokenType == 'function_def' && token.hasOwnProperty('return') && token.return != null && token.hasOwnProperty('returnKeywords')){
                for(const funcReturn of token.return){
                    if (httpLibs.includes(funcReturn)){
                        for(const keyword of token.returnKeywords){
                            if (keyword[0] == 'verify' && keyword[1] == false) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG_ON_RETURN);
                        }
                    }
                }
            }
        } catch(error){
            console.log(error);
        }
    },
    
    triggerAlarm: (fileName, MSG, lineno, WARNING_MSG) => {
        let backslashSplittedFilePathLength = fileName.split("/").length
        let filenameFromPath = fileName.split("/")[backslashSplittedFilePathLength - 1]
        
        vscode.window.showWarningMessage(MSG +" : "+ filenameFromPath+":"+lineno);
        console.log( "\u001b[1;31m"+"warning: "+MSG +"  location:"+ fileName+":"+lineno);
        fs.appendFileSync('smell-spotter/warning-logs/project_warnings.csv', fileName+" ,"+WARNING_MSG+"\n");
    }
}

module.exports = smell;

File: py-scripts/filters/nocertificate.py
class NoCertificate:
    '''This is the class for detecting no certificate validations in code'''

    def __init__(self):
        self.warning_message = 'no certificate validation'
        self.http_libs = ['requests.get','requests.Session.get', 'requests.post', 'requests.Session.get', 'requests.head']
        self.context_variables = ['requests.Session.verify']
    

    def detect_smell(self, token, src_file):
        try:
            if token.__contains__("line"): lineno = token["line"]
            if token.__contains__("type"): tokenType = token["type"]
            if token.__contains__("name"): name = token["name"]
                
            if tokenType == "variable" and name in self.context_variables and token['value'] is False: 
                self.trigger_alarm(src_file, lineno)

            elif tokenType == "variable" and token.__contains__("valueSrc") and token.__contains__("funcKeywords"):
                keywords = token['funcKeywords']
                valueSrc = token['valueSrc']

                if (valueSrc in self.http_libs or self.is_http_call_relaxed(valueSrc)) and len(keywords) > 0:
                    
                    for keyword in keywords:
                        if keyword[0] == 'verify' and (keyword[1] is False or keyword[1] == 'False'): 
                            self.trigger_alarm(src_file, lineno)

            elif tokenType == "function_call" and (name in self.http_libs or self.is_http_call_relaxed(name)) and token.__contains__("keywords"):
                keywords = token["keywords"]
                
                for keyword in keywords:
                    if keyword[0] == 'verify' and (keyword[1] is False or keyword[1] == 'False'): 
                        self.trigger_alarm(src_file, lineno)
            
            
            elif tokenType == "function_def" and token.__contains__("return") and token.__contains__("returnKeywords") and token["return"] is not None:
                
                keywords = token['returnKeywords']
                
                for func_return in token['return']:
                    if func_return in self.http_libs or self.is_http_call_relaxed(func_return):
                        for keyword in keywords:
                            if keyword[0] == 'verify' and (keyword[1] is False or keyword[1] == 'False'): 
                                self.trigger_alarm(src_file, lineno)
                
            
        except Exception as error: 
            print(str(error))


    def is_http_call_relaxed(self, method_name):
        self.http_libs_verb = ['.get', '.post']

        if isinstance(method_name, str) is False: return False

        for lib in self.http_libs:
            if lib in method_name: 
                return True
        
        for lib_verb in self.http_libs_verb:
            if lib_verb in method_name and 'request' in method_name.lower().strip():
                return True
        
        return False


    def trigger_alarm(self, src_file, lineno):
        print(src_file +":"+ str(lineno)+" ,"+self.warning_message)
    